segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/FixedKeyLengthTableSegmentLayout.java

```diff
@@ -0,0 +1,534 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+/**
+ * Fixed-Key-Length Table Segment layout.
+ *
+ * This Layout supports only fixed-length keys (declared at segment creation), but is significantly less complex and
+ * resource intensive than {@link HashTableSegmentLayout} and provides order (sorting) between Keys.
+ */
+@Beta
+@Slf4j
+class FixedKeyLengthTableSegmentLayout extends TableSegmentLayout {
+    private final DelayedProcessor<CompactionCandidate> compactionService;
+    private final TableCompactor.Config tableCompactorConfig;
+
+    //region Constructor
+
+    FixedKeyLengthTableSegmentLayout(Connector connector, TableExtensionConfig config, ScheduledExecutorService executor) {
+        super(connector, config, executor);
+        this.compactionService = new DelayedProcessor<>(this::compactIfNeeded, config.getCompactionFrequency(), executor,
+                String.format("TableCompactor[%s]", connector.getContainerId()));
+        this.tableCompactorConfig = new TableCompactor.Config(config.getMaxCompactionSize());
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    public void close() {
+        this.compactionService.close();
+        log.info("{}: Closed.", this.traceObjectId);
+    }
+
+    //endregion
+
+    //region TableSegmentLayout Implementation
+
+    @Override
+    Collection<WriterSegmentProcessor> createWriterSegmentProcessors(UpdateableSegmentMetadata metadata) {
+        // We don't need any Writer Processors.
+        return Collections.emptyList();
+    }
+
+    @Override
+    Map<AttributeId, Long> getNewSegmentAttributes(@NonNull TableSegmentConfig config) {
+        ensureValidKeyLength("", config.getKeyLength());
+        Preconditions.checkArgument(config.getKeyLength() % Long.BYTES == 0, "KeyLength must be a multiple of %s; given %s.", Long.BYTES, config.getKeyLength());
+        val result = new HashMap<AttributeId, Long>();
+        result.put(Attributes.ATTRIBUTE_ID_LENGTH, (long) config.getKeyLength());
+        result.putAll(this.config.getDefaultCompactionAttributes());
+        return result;
+    }
+
+    @Override
+    CompletableFuture<Void> deleteSegment(@NonNull String segmentName, boolean mustBeEmpty, Duration timeout) {
+        if (mustBeEmpty) {
+            throw new UnsupportedOperationException("mustBeEmpty not supported on Fixed-Key-Length Table Segments.");
+        }
+
+        return this.connector.deleteSegment(segmentName, timeout)
+                .thenRun(() -> this.compactionService.cancel(segmentName));
```

Comment:
The compaction service will only run one item at a time (so it won't do tasks in parallel - so no chance of overloading the segment store). It also only treats segments as units. Even if a segment may be "queued" multiple times in short sequence, it only considers it once; and as soon as its compaction finishes, that segment won't be eligible for compaction (due to less fragmentation) for a while. Plus there is a cool-down period (one segment may not be reprocessed more frequently than X minutes).

Based on these, I do not think this will add any visible load to the Segment store. If we are concerned about it building a large queue, then that may be, but we will need to have a large number of active table segments to work on. It will be a function of load, so if we get there, the whole system will be busy, not just this. If we get there, we can consider tweaking the settings to properly balance these.

The settings involved are the cool-down period, parallelism, compaction thresholds, etc.